src/main/scala/com/wavesplatform/database/Caches.scala

```diff
@@ -92,21 +92,47 @@ trait Caches extends Blockchain with ScorexLogging {
     }
   }
 
-  private val transactionIds                                       = new util.HashMap[ByteStr, Long]()
-  protected def forgetTransaction(id: ByteStr): Unit               = transactionIds.remove(id)
-  override def containsTransaction(id: ByteStr): Boolean           = transactionIds.containsKey(id)
-  override def learnTransactions(values: Map[ByteStr, Long]): Unit = transactionIds.putAll(values.asJava)
-  override def forgetTransactions(pred: (ByteStr, Long) => Boolean): Map[ByteStr, Long] = {
-    val removedTransactions = Map.newBuilder[ByteStr, Long]
-    val iterator            = transactionIds.entrySet().iterator()
-    while (iterator.hasNext) {
-      val e = iterator.next()
-      if (pred(e.getKey, e.getValue)) {
-        removedTransactions += e.getKey -> e.getValue
-        iterator.remove()
+  protected def rememberBlocksInterval: Long
+
+  private val blocksTs                               = new util.HashMap[Int, Long] // Height -> block timestamp
+  private var olderStoredBlockTimestamp              = Long.MaxValue
+  private val transactionIds                         = new util.HashMap[ByteStr, Int]() // TransactionId -> height
+  protected def forgetTransaction(id: ByteStr): Unit = transactionIds.remove(id)
+  private var miss: Int                              = 0
+  override def containsTransaction(tx: Transaction): Boolean = transactionIds.containsKey(tx.id()) || {
+    if (tx.timestamp - 120 * 60 * 1000 <= olderStoredBlockTimestamp) {
+      miss += 1
+      log.info(s"Cache miss, do lookup (total $miss)")
+      transactionHeight(tx.id()).nonEmpty
+    } else {
+      false
+    }
+  }
+  def forgetBlocks(): Unit = {
+    var olderBlock = Int.MaxValue;
+    {
+      val biterator = blocksTs.entrySet().iterator()
```

Comment:
`blockTs` should likely be `java.util.TreeMap`, so that:
1. you could just use `removeIf(_.getKey < blockTimestamp - rememberBlocksInterval)`
1. `oldestBlockTimestamp = blockTs.values().next()`